src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java

```diff
@@ -0,0 +1,909 @@
+import com.google.firebase.ImplFirebaseTrampolines;
+import com.google.firebase.internal.CallableOperation;
+import com.google.firebase.internal.FirebaseRequestInitializer;
+import com.google.firebase.internal.FirebaseScheduledExecutor;
+import com.google.firebase.internal.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+class FirebaseProjectManagementServiceImpl implements AndroidAppService, IosAppService {
+  @VisibleForTesting static final String FIREBASE_PROJECT_MANAGEMENT_URL =
+      "https://firebase.googleapis.com";
+  @VisibleForTesting static final int MAXIMUM_LIST_APPS_PAGE_SIZE = 100;
+  @VisibleForTesting static final String PATCH_OVERRIDE_KEY = "X-HTTP-Method-Override";
+  @VisibleForTesting static final String PATCH_OVERRIDE_VALUE = "PATCH";
+
+  private static final int MAXIMUM_POLLING_ATTEMPTS = 8;
+  private static final long POLL_BASE_WAIT_TIME_MILLIS = 500L;
+  private static final double POLL_EXPONENTIAL_BACKOFF_FACTOR = 1.5;
+  private static final String ANDROID_APPS_RESOURCE_NAME = "androidApps";
+  private static final String IOS_APPS_RESOURCE_NAME = "iosApps";
+  private static final String ANDROID_NAMESPACE_PROPERTY = "package_name";
+  private static final String IOS_NAMESPACE_PROPERTY = "bundle_id";
+  private static final ImmutableMap<Integer, String> ERROR_CODES =
+      ImmutableMap.<Integer, String>builder()
+          .put(401, "Request not authorized.")
+          .put(403, "Client does not have sufficient privileges.")
+          .put(404, "Failed to find the resource.")
+          .put(409, "The resource already exists.")
+          .put(429, "Request throttled by the backend server.")
+          .put(500, "Internal server error.")
+          .put(503, "Backend servers are over capacity. Try again later.")
+          .build();
+
+  private final FirebaseApp app;
+  private final JsonFactory jsonFactory;
+  private final HttpRequestFactory requestFactory;
+  private final ScheduledExecutorService scheduledExecutor;
+
+  private final CreateAndroidAppFromAppIdFunction createAndroidAppFromAppIdFunction =
+      new CreateAndroidAppFromAppIdFunction();
+  private final CreateIosAppFromAppIdFunction createIosAppFromAppIdFunction =
+      new CreateIosAppFromAppIdFunction();
+
+  private HttpResponseInterceptor interceptor;
+
+  FirebaseProjectManagementServiceImpl(FirebaseApp app) {
+    this.app = app;
+
+    JsonFactory jsonFactory = app.getOptions().getJsonFactory();
+    this.jsonFactory = jsonFactory;
+
+    HttpTransport httpTransport = app.getOptions().getHttpTransport();
+    this.requestFactory =
+        httpTransport.createRequestFactory(new FirebaseRequestInitializer(app));
+
+    this.scheduledExecutor = new FirebaseScheduledExecutor(
```

Comment:
Please remove this. No service should initialize its own thread pools in the SDK.

Reply:
As we discussed, I have re-implemented the blocking API with `Thread.sleep()`, and instead of exposing the entire `ScheduledExecutorService` from the `FirebaseApp`, I have instead exposed three separate methods `transform`, `transformAsync`, and `schedule`.